File: 백준/Silver/1149. RGB거리/RGB거리.py
import sys

input = sys.stdin.readline

N = int(input())
dp = [[sys.maxsize] * 3 for _ in range(N)]

RGB = []
for i in range(N):
    RGB.append(list(map(int, input().split())))

dp[0][0] = RGB[0][0]
dp[0][1] = RGB[0][1]
dp[0][2] = RGB[0][2]

for i in range(1, N):
    dp[i][0] = min(dp[i - 1][1], dp[i - 1][2]) + RGB[i][0]
    dp[i][1] = min(dp[i - 1][0], dp[i - 1][2]) + RGB[i][1]
    dp[i][2] = min(dp[i - 1][0], dp[i - 1][1]) + RGB[i][2]

print(min(dp[N-1]))

File: 백준/Silver/1149. RGB거리/RGB거리.java
import java.io.*;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {

	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static StringTokenizer st;

	public static void main(String[] args) throws IOException {
		// RGB 거리

		int N = Integer.parseInt(br.readLine());

		int[] dp = new int[3]; // 이전

		for (int i = 0; i < N; i++) {
			st = new StringTokenizer(br.readLine());
			if (i == 0) {
				dp[0] = Integer.parseInt(st.nextToken());
				dp[1] = Integer.parseInt(st.nextToken());
				dp[2] = Integer.parseInt(st.nextToken());
				continue;
			}
			
			int[] next = new int[3];
			int a = Integer.parseInt(st.nextToken());
			int b = Integer.parseInt(st.nextToken());
			int c = Integer.parseInt(st.nextToken());

			// 첫번째
			next[0] = Math.min(dp[1] + a, dp[2] + a);
			next[1] = Math.min(dp[0] + b, dp[2] + b);
			next[2] = Math.min(dp[0] + c, dp[1] + c);
			
			dp = next;
		}
		int minD = dp[0];
		minD = Math.min(minD,dp[1]);
		minD = Math.min(minD,dp[2]);
		System.out.println(minD);
	}

}
